### PR TITLE
Add CI, ESLint, and Travis

### PR DESCRIPTION
**Pull request by gcochard**
- Adding travis config
- Add pre-commit hook to lint and pre-push to test
- Add mocha tests
- Move to window.<name> everywhere.
  - This is necessary for tests to work properly, and does not cause any
    regressions in the browser. Tests currently use mocks for Chrome.*
    methods and localstorage.

### CODE REVIEW

**Comment by gcochard** [comment]
@attilaolah this is basically done, albeit very much later than I had hoped. Let me know if you need any guidance on the tests, but the framework is there and fairly easy to use if you know the [Sinon API](http://sinonjs.org/docs/).

**Comment by PeterDaveHello** [comment]
:+1:

File: package.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "github-notifier-chrome",
+  "version": "2.0.0",
+  "description": "> Chrome/Opera extension - Displays your GitHub notifications unread count",
+  "main": "index.js",
+  "scripts": {
+    "lint": "eslint extension/*.js",
```
Comment (attilaolah) [inline package.json]: Why not lint `test/*.js` as well, while there? I know it doesn't matter, but it should help keep the repo clean overall.

**Comment by attilaolah** [comment]
LGTM.

@gcochard who owns the project on Travis?

Also, is it possible to get Travis to push commit statuses to GitHub? Or will that start working once this is merged?

**Comment by arthurvr** [comment]
I'd rather wait with setting up ESLint as @sindresorhus was planning to add it to all this repos. I'd rather have a consistent setup between all repos.

File: .eslintrc [hunk 2]
```diff
@@ -0,0 +1,32 @@
+{
+    "env": {"node": true},
```
Comment (arthurvr) [inline .eslintrc]: wrong indentation. We use tabs.

File: .eslintrc [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    "globals": {},
```
Comment (arthurvr) [inline .eslintrc]: No deal with having this when it's empty? Or is there?

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+node_modules/
+npm-debug.log
```
Comment (arthurvr) [inline .gitignore]: npm-debug.log should be in your global gitignore, not here.

File: .travis.yml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+language: node_js
+
+node_js:
+  - "0.12"
```
Comment (arthurvr) [inline .travis.yml]: single-quotes

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+language: node_js
+
```
Comment (arthurvr) [inline .travis.yml]: remove this line

File: .travis.yml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+language: node_js
```
Comment (arthurvr) [inline .travis.yml]: Please also enable the container based-infrastructure. (http://docs.travis-ci.com/user/workers/container-based-infrastructure/)

File: package.json [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    "github",
+    "notifier",
+    "chrome"
+  ],
+  "author": "Sindre Sorhus",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/sindresorhus/github-notifier-chrome/issues"
+  },
+  "homepage": "https://github.com/sindresorhus/github-notifier-chrome#readme",
```
Comment (arthurvr) [inline package.json]: moot

File: package.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    "url": "git+https://github.com/sindresorhus/github-notifier-chrome.git"
+  },
+  "keywords": [
+    "github",
+    "notifier",
+    "chrome"
+  ],
+  "author": "Sindre Sorhus",
+  "license": "MIT",
+  "bugs": {
```
Comment (arthurvr) [inline package.json]: moot if you have `repo`

File: package.json [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/sindresorhus/github-notifier-chrome.git"
+  },
+  "keywords": [
+    "github",
+    "notifier",
+    "chrome"
+  ],
+  "author": "Sindre Sorhus",
```
Comment (arthurvr) [inline package.json]: Follow this format: https://github.com/sindresorhus/time-grunt/blob/master/package.json#L18.

File: package.json [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+  "main": "index.js",
+  "scripts": {
+    "lint": "eslint extension/*.js",
+    "test": "mocha"
+  },
+  "pre-commit": [
+    "lint",
+    "test"
+  ],
+  "repository": {
```
Comment (arthurvr) [inline package.json]: should be enough as `"sindresorhus/github-notifier-chrome"`

File: test/main.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    hash: "",
+    host: "",
+    hostname: "",
+    href: "",
+    origin: "",
+    pathname: "",
+    port: "",
+    protocol: "",
+    search: ""
+  }
```
Comment (arthurvr) [inline test/main.js]: wrong indentation.

Comment (arthurvr) [inline test/main.js]: same for the test of this file.

File: test/main.js [hunk 13]
```diff
@@ -0,0 +1,66 @@
+'use strict';
+var assert = require('assert');
+var sinon = require('sinon');
+global.window = {
+  location:{
```
Comment (arthurvr) [inline test/main.js]: Can you double-check this whole file for code-style violations? We use single-quotes, tabs, space after keywords, ...

File: test/main.js [hunk 14]
```diff
@@ -0,0 +1,66 @@
+'use strict';
+var assert = require('assert');
+var sinon = require('sinon');
```
Comment (arthurvr) [inline test/main.js]: leave an empty line after this.

File: test/main.js [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+XMLHttpRequest.prototype.getHeader = function getHeader(name){
+    return this.responseHeaders[name];
+};
+
+global.XMLHttpRequest = XMLHttpRequest;
+var chromeStub = require('chrome-stub');
+var localMock = new require('mock-localstorage')();
+global.localStorage = localMock;
+global.chrome = chromeStub;
+describe('main',function(){
```
Comment (arthurvr) [inline test/main.js]: If you just describe `main` then don't do the `describe` at all.

**Comment by arthurvr** [comment]
Neat work. Left some comments :dancer: 

> LGTM.

@attilaolah don't lgtm too fast ;)

File: package.json [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    "type": "git",
+    "url": "git+https://github.com/sindresorhus/github-notifier-chrome.git"
+  },
+  "keywords": [
+    "github",
+    "notifier",
+    "chrome"
+  ],
```
Comment (arthurvr) [inline package.json]: Some more keywords we can come up with?

File: package.json [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+  "description": "> Chrome/Opera extension - Displays your GitHub notifications unread count",
+  "main": "index.js",
+  "scripts": {
+    "lint": "eslint extension/*.js",
+    "test": "mocha"
+  },
+  "pre-commit": [
+    "lint",
+    "test"
+  ],
```
Comment (arthurvr) [inline package.json]: I'd rather not do a pre-commit hook.

File: package.json [hunk 18]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "github-notifier-chrome",
+  "version": "2.0.0",
+  "description": "> Chrome/Opera extension - Displays your GitHub notifications unread count",
```
Comment (arthurvr) [inline package.json]: `> `?

File: test/main.js [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+  it('should register alarm callback',function(){
+    require('../extension/api');
+    require('../extension/main');
+    assert(chromeStub.alarms.create.called);
+  });
+  it('should do something on alarm fired',function(done){
```
Comment (arthurvr) [inline test/main.js]: Keep a line between two `it` blocks.

File: test/main.js [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    this.requestHeaders[name] = val;
+};
+
+XMLHttpRequest.prototype.getHeader = function getHeader(name){
+    return this.responseHeaders[name];
+};
+
+global.XMLHttpRequest = XMLHttpRequest;
+var chromeStub = require('chrome-stub');
+var localMock = new require('mock-localstorage')();
```
Comment (arthurvr) [inline test/main.js]: `new require`?

Comment (arthurvr) [inline test/main.js]: Just require this on top of the file.

File: test/main.js [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+XMLHttpRequest.prototype.setRequestHeader = function setRequestHeader(name,val){
+    this.requestHeaders[name] = val;
+};
+
+XMLHttpRequest.prototype.getHeader = function getHeader(name){
+    return this.responseHeaders[name];
+};
+
+global.XMLHttpRequest = XMLHttpRequest;
+var chromeStub = require('chrome-stub');
```
Comment (arthurvr) [inline test/main.js]: Just require this on top of the file.

File: .gitignore [hunk 22]
```diff
@@ -0,0 +1,2 @@
+node_modules/
```
Comment (arthurvr) [inline .gitignore]: Remove the `/`

**Comment by attilaolah** [comment]
> > LGTM.

> @attilaolah don't lgtm too fast ;)

Actually I entirely ignored the metafiles like `package.json` and the ESLint settings. I figured, if it runs the linter, it's good enough for me — but good thing we have someone on board to enforce strictness, @arthurvr :bowtie:

**Comment by gcochard** [comment]
@arthurvr @attilaolah I think @sindresorhus has to turn on travis for the repo before it starts working. I have [one for my fork](https://travis-ci.org/gcochard/github-notifier-chrome) that I set up for testing these changes before making the PR.

RE: Eslint config, I'm good with waiting until all repos are added, we should probably just use a shared config. Where is that being discussed?

File: package.json [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+  "license": "MIT",
+  "devDependencies": {
+    "chrome-stub": "^1.2.0",
+    "eslint": "^0.20.0",
+    "istanbul": "^0.3.13",
+    "mocha": "^2.2.4",
+    "mock-localstorage": "^0.1.3",
+    "sinon": "^1.15.4"
+  }
+}
```
Comment (sindresorhus) [inline package.json]: No point in filling all of this out. It's not going on npm anyways. It's enough with `devDependencies` and `scripts`.

**Comment by sindresorhus** [comment]
Travis turned on.

**Comment by gcochard** [comment]
@arthurvr does this look good to you now?

Would everyone prefer a single squashed commit?

File: extension/api.js [hunk 24]
```diff
@@ -58,14 +58,14 @@
 	})();
 
 	window.gitHubNotifCount = function (cb) {
-		var token = GitHubNotify.settings.get('oauthToken');
+		var token = window.GitHubNotify.settings.get('oauthToken');
```
Comment (sindresorhus) [inline extension/api.js]: Why? I don't see the point of this.

File: package.json [hunk 25]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "github-notifier-chrome",
```
Comment (sindresorhus) [inline package.json]: name is also moot

**Comment by sindresorhus** [comment]
Almost. A few super minor inline comments.

File: package.json (hunk 25 above)
Comment (gcochard) [inline package.json]: For `npm` to parse the file, I believe it needs a `name` field. I will check later tonight, on my phone now.

File: extension/api.js (hunk 24 above)
Comment (gcochard) [inline extension/api.js]: The only reason I did this is for the tests. It does not affect the behavior when running in chrome, but mocha/node need the `window.` prepended.

File: package.json (hunk 25 above)
Comment (gcochard) [inline package.json]: This may have been fixed in npm 2 or 3, so that may be why I have this here.

**Comment by attilaolah** [comment]
> Would everyone prefer a single squashed commit?

I'm not a fan of squashing commits, unless you're changing binary files or include commits reverting each other. I wouldn't bother squashing.

**Comment by sindresorhus** [comment]
> I'm not a fan of squashing commits, unless you're changing binary files or include commits reverting each other. I wouldn't bother squashing.

Multiple commits in a PR is ok as I can clearly see the changes, but I squash on merge to keep the master commit log clean and focused.

File: extension/api.js (hunk 24 above)
Comment (sindresorhus) [inline extension/api.js]: You can just add it as a global using the ESLint comment directive. http://eslint.org/docs/user-guide/configuring.html

File: test/index.js [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+	this.url = url;
+};
+
+XMLHttpRequest.prototype.send = function send() {
+	this.readyState = 4;
+	this.responseText = '[]';
+	this.status = 200;
+	this.responseHeaders['X-Poll-Interval'] = '60';
+	if (this.onreadystatechange) {
+	setTimeout(this.onreadystatechange, 50);
```
Comment (arthurvr) [inline test/index.js]: missing indentation

Comment (attilaolah) [inline test/index.js]: Exactly why I think the linter could have a look at the test files as well.

Comment (arthurvr) [inline test/index.js]: :+1:

File: .travis.yml [hunk 27]
```diff
@@ -0,0 +1,5 @@
...
+node_js:
+  - '0.12'
+sudo: false
```
Comment (arthurvr) [inline .travis.yml]: move this one to the top :)

**Comment by sindresorhus** [comment]
Landed. Good stuff @gcochard :)

**Comment by gcochard** [comment]
Awesome :+1:

**Pull request closed without merging**